### PR TITLE
Toward better implicit resolution

### PR DESCRIPTION
**Pull request by raulraja**
The following PR removes the need to register global typeclass instances manually by delegating initial lookup to naming conventions.

If you declare a typeclass instance with a companion and a function called `instance` inside all the work is done for you. Also if you use `@deriving` there is nothing further that needs to happen
```kotlin
interface OptionFunctorInstance : Functor<OptionHK> {
  ...
}

object OptionFunctorInstanceImplicits {
    fun instance(): OptionFunctorInstance = object : OptionFunctorInstance {}
}
```
You can now safely invoke:
```kotlin
functor<OptionHK>() 
```
The instance is loaded from a class  that should be named `[Option][Functor]Instance` in this case. The lookup function uses the companion `instance` function to resolve the instance with reflection the first time and it caches it for subsequent calls.

If you do not wish to call the class that name you can still use manual registration with `registerInstance`
```kotlin
registerInstance(myOptionFunctorInstance, Functor::class, Option::class, ...)
```
Typeclass instances no longer recurse on their interfaces for implementation so we are enforcing now proper separation of interfaces in the declaration hierarchy to avoid instances getting registered automatically for subtypes.

Initial code is also provided to support typeclasses with arity > 1 once we support derivation for those.

### CODE REVIEW

File: kategory-annotations/build.gradle [hunk 1]
```diff
@@ -2,6 +2,7 @@ apply plugin: 'kotlin'
 
 dependencies {
     compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    compile "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
```
Comment (JorgeCastilloPrz) [inline kategory-annotations/build.gradle]: how big is the kotlin reflection lib?

File: kategory-annotations/src/main/java/kategory/Typeclass.kt [hunk 2]
```diff
@@ -127,13 +79,70 @@ data class TypeClassInstanceNotFound(val type: Type)
...
+ * Obtains a global registered typeclass instance when fast unsafe runtime lookups are desired over passing instances
+ * as args to functions. Use with care. This lookup will throw an exception if a typeclass is summoned and can't be found
+ * in the GlobalInstances map or the companion factory methods for the `on` target type
+ */
 @Suppress("UNCHECKED_CAST")
-fun <T : Typeclass> instance(t: Type): T {
-    if (GlobalInstances.containsKey(t))
-        return GlobalInstances.getValue(t) as T
+fun <T : Typeclass> instance(of: KClass<T>, on: KClass<*>, vararg typeArgs: KClass<*>): T =
+    instanceImpl(of, on, *typeArgs) as T
```
Comment (pakoito) [inline kategory-annotations/src/main/java/kategory/Typeclass.kt]: To maintain type safety, we know that Typeclass will be parametrized at least on a witness type F. If typeclass would extend HK<TypeclassWitness, F>, then we could return HK<TypeclassWitness, F> here and make the user provide the evidence to cast it to the correct type.

File: kategory-annotations/build.gradle (hunk 1 above)
Comment (pakoito) [reply kategory-annotations/build.gradle]: It is quite large, IIRC.

Comment (pakoito) [reply kategory-annotations/build.gradle]: EDIT: It says here it's 2MB O.o

https://github.com/Kotlin/kotlinx.reflect.lite

Comment (nomisRev) [reply kategory-annotations/build.gradle]: > The size of the minimized version of this library is about 100Kb plus the dependency on protobuf-java

IIRC it is so big due to proto buff library. I'm not sure if the lite version would suffice for us but 100kb+ might be justifiable for our use-case ?

File: kategory-annotations/src/main/java/kategory/Typeclass.kt (hunk 2 above)
Comment (pakoito) [reply kategory-annotations/src/main/java/kategory/Typeclass.kt]: EDIT: Nevermind, a type parameter would be as unsafe as this implementation anyway.

**Comment by raulraja** [comment]
@JorgeCastilloPrz yes, we would have to write different interfaces which is good practice because users may want to use those separately for their own composition. The file where they are located should not matter since they are loaded by fully qualified names including package.

File: kategory-annotations/build.gradle (hunk 1 above)
Comment (raulraja) [reply kategory-annotations/build.gradle]: I didn't now about the size. The lite version linked there does not have `companionObjectInstace`. I'm gonna try to rewrite it with java reflection. Can't wait for the @implicit support to get rid of all this shit. xD

**Comment by raulraja** [comment]
@pakoito @JorgeCastilloPrz @nomisRev reworked the code to not depend on Kotlin reflect

**Pull request merged**